File: ascii.rb
require "mini_magick"

print "Enter file name: "
filename = gets.chomp

image = MiniMagick::Image.open(filename)
characters = "`^\",:;Il!i~+_-?][}{1)(|\\/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$"

pixels = image.get_pixels
for i in 0..(pixels.length - 1)
    for j in 0..(pixels[i].length - 1)
        rgb = pixels[i][j]
        brightness = (rgb.reduce :+) / 3
        index = (brightness.to_f / 255 * 65).round
        if index == 0
            print " "
        else
            print characters[index]
        end
    end
    print "\n"
end

File: ascii.py
from PIL import Image
import time

filename = input("Which file to use: ")

before = time.time()

CHARACTERS = '`^",:;Il!i~+_-?][}{1)(|\\/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$'
N_CHARACTERS = len(CHARACTERS) - 1

image = Image.open(filename)
png = filename.endswith(".png")

for y in range(image.height):
    for x in range(image.width):
        rgb = image.getpixel((x, y))
        if png:
            *rgb, a = rgb
        brightness = sum(rgb) / 3
        if brightness == 0:
            print(" ", end="")
            continue
        print(CHARACTERS[round(N_CHARACTERS / 255 * brightness)], end="")
    print("")

after = time.time()
print(f"Took {after - before:.2}s")
